### PR TITLE
Don't do version neg on an HRR

### LINKED ISSUE #4292: TLS 1.3 Retry Client Hello does not follow spec
I tested TLS 1.3 s_client against a server which required a different keyshare than the default offered.  The second client hello is missing extensions offered in the first client hello and the random value changes.

There was a recent discussion on this topic on the IETF TLS mailing list  and the conclusion was random should not change and only certain extensions may be altered.  https://www.ietf.org/mail-archive/web/tls/current/msg24409.html

Attached is a pcap.  (I zipped the file because github said pcap was not supported)
[s_client_hrr.pcap.zip](https://github.com/openssl/openssl/files/1264050/s_client_hrr.pcap.zip)

### PR DESCRIPTION
**Pull request by mattcaswell**
Previously if a client received an HRR then we would do version negotiation
immediately - because we know we are going to get TLSv1.3. However this
causes a problem when we emit the 2nd ClientHello because we start changing
a whole load of stuff to ommit things that aren't relevant for < TLSv1.3.
The spec requires that the 2nd ClientHello is the same except for changes
required from the HRR. Therefore the simplest thing to do is to defer the
version negotiation until we receive the ServerHello.

Fixes #4292

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

**Comment by t-j-h** [comment]
This looks like a nice clean solution to me.

**Comment by mattcaswell** [comment]
Realised that although we're not doing version neg on the HRR, we should still sanity check the version field. Added a commit for that.

File: ssl/record/ssl3_record.c [hunk 1]
```diff
@@ -271,8 +271,13 @@ int ssl3_get_record(SSL *s)
...
                 thisrr->rec_version = version;
 
-                /* Lets check version. In TLSv1.3 we ignore this field */
+                /*
+                 * Lets check version. In TLSv1.3 we ignore this field. For an
+                 * HRR we haven't actually selected TLSv1.3 yet, but we still
+                 * treat it as TLSv1.3, so we must check for that explicitly
+                 */
                 if (!s->first_packet && !SSL_IS_TLS13(s)
+                        && !s->hello_retry_request
```
Comment (kaduk) [inline ssl/record/ssl3_record.c]: For the actual HelloRetryRequest record, ```s->first_packet``` should be set, right?
So the effect of checking ```s->hello_retry_request``` is really for when we are processing the ServerHello, which is in this flow the second record we're reading?

Comment (mattcaswell) [reply ssl/record/ssl3_record.c]: Right! Good point. I've tweaked the comment wording accordingly.

File: ssl/record/ssl3_record.c [hunk 2]
```diff
@@ -272,9 +272,10 @@ int ssl3_get_record(SSL *s)
                 thisrr->rec_version = version;
 
                 /*
-                 * Lets check version. In TLSv1.3 we ignore this field. For an
-                 * HRR we haven't actually selected TLSv1.3 yet, but we still
-                 * treat it as TLSv1.3, so we must check for that explicitly
+                 * Lets check version. In TLSv1.3 we ignore this field. For the
+                 * ServerHello after an HRR we haven't actually selected TLSv1.3
+                 * yet, but we stilltreat it as TLSv1.3, so we must check for
```
Comment (kaduk) [inline ssl/record/ssl3_record.c]: You might add the space in "still treat" before merging.

Comment (mattcaswell) [reply ssl/record/ssl3_record.c]: Fixed

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**